packages/web-components/fast-element/src/interfaces.ts

```diff
@@ -90,17 +90,65 @@ export interface FASTGlobal {
     addMessages(messages: Record<number, string>): void;
 }
 
+let kernelMode;
+const kernelAttr = "fast-kernel";
+
+try {
+    if (document.currentScript) {
+        kernelMode = document.currentScript.getAttribute(kernelAttr);
+    } else {
+        const scripts = document.getElementsByTagName("script");
+        const currentScript = scripts[scripts.length - 1];
+        kernelMode = currentScript.getAttribute(kernelAttr);
+    }
+} catch (e) {
+    kernelMode = "isolate";
+}
+
 /**
- * Core services shared across FAST instances.
+ * Core services that can be shared across FAST instances.
  * @internal
  */
-export const enum KernelServiceId {
-    updateQueue = 1,
-    observable = 2,
-    contextEvent = 3,
-    elementRegistry = 4,
+type KernelServiceId = {
+    readonly updateQueue: string | number;
+    readonly observable: string | number;
+    readonly contextEvent: string | number;
+    readonly elementRegistry: string | number;
+};
+
+let KernelServiceId: KernelServiceId;
+
+switch (kernelMode) {
+    case "share": // share the kernel across major versions
+        KernelServiceId = Object.freeze({
+            updateQueue: 1,
+            observable: 2,
+            contextEvent: 3,
+            elementRegistry: 4,
+        });
+        break;
+    case "share-v2": // only share the kernel with other v2 instances
+        KernelServiceId = Object.freeze({
+            updateQueue: 1.2,
+            observable: 2.2,
+            contextEvent: 3.2,
+            elementRegistry: 4.2,
+        });
+        break;
+    default:
+        // fully isolate the kernel from all other FAST instances
+        const postfix = `-${Math.random().toString(36).substring(2, 8)}`;
```

Comment:
The `share-v2` mode will share the kernel between instances of the same v2 major version package. The default is that every instance of fast is now isolated. For that we need a random identifier for each service otherwise we could end up with sharing on accident.

Are you suggesting that `share-v2` should be the default rather than `isloate`?

Reply:
It seems like we only want to make it different per version (V1/V2/V3)? Can we get FAST version and use it as the prefix for the names, instead of using random number which may produce something which is hard to control? -- if the version resolving tool has issue, we may end up with a large number of kernels. Can we only patch it like `updateQueue: '${fast-version}-1'`? And then V1 code will still work because they are sharing the same kernel, and V2 code will work in a different kernel?

Reply:
I think there really are three scenarios that are important:

1. Share the kernel across both v1 and v2.
2. Let v1 instances share their kernel and let v2 instances share their kernel.
3. Ensure that the v2 kernel is isolated and not shared even among other v2 instances.

We cannot change how the v1 kernel is shared. We can change how v2's kernel is shared. This PR adds the three modes described above and proposes to make "no sharing" the default. We could make "sharing only with other v2 instances" the default. But we cannot make sharing across v1 and v2 the default. That has to be opt-in. Do we want to change the default?

I do think we need all of these options. The random number code just ensures that the service ids are unique so that they can't be shared at all for mode 3 (by design).

Reply:
Thanks to explain, I think #3 is what I got confused: what's the use case that we want V2 kernel to not shared to the other V2 instances?

Reply:
That is the way that v1 originally worked before we introduced sharing. So, I would like to bring back the original behavior as an option.